### DISCUSSION TITLE
categorical variables

### DISCUSSION THREAD

**Discussion by mturk24** [ourownstory/neural_prophet#139] [Feature Requests & Feedback]
Hi there, if I am trying to forecast a categorical variable "y" that has values 1 and 0 (1 = take action, 0 = don't take action) with just the time series "ds" and "y" as input, I'm getting an error when observing my model fitted results on default parameters for both the neural prophet object and my fit method call.

My fit method looks like the following:
model.fit(X_train, freq='D')

However, my output looks like the below

<img width="591" alt="Screen Shot 2020-12-01 at 11 33 15 PM" src="https://user-images.githubusercontent.com/8604219/100828665-97ce8980-342d-11eb-9330-88b7dc53f7db.png">

Would you be able to provide a simple example of fitting a model on a categorical variable?

**Comment by mturk24** [comment]
Moreover, when are you estimating multi-variable forecasts to be available?

**Comment by ourownstory** [comment]
Hi Matthew,

Thank you for reporting this issue. 

Officially, the model does not support categorical variables, as that would not work with concepts such as tend and seasonality in their original sense. 

In your case, as you only have one category, you should be able to model it as a binary series. Try if setting `normalize="minmax"` resolves your issue?

Would you mind providing some example data and a minimal code example for us to reproduce the issue?

Re: Multi-Variable forecasts: While it is something we plan to support, we currently do not have a planned date, as it would require a restructuring of larger parts of the API, particularly for plotting. However, If you would like to implement it, I would be happy to chat with you to see how we could best achieve that.

**Comment by mturk24** [comment]
@ourownstory Hi thanks for getting back to me! 

By not supporting categorical variables, what do you mean exactly? Given your suggestion to model a single category as a binary series, do you just mean not able to support multiple categorical variables (as opposed to one as the the target variable to predict or even as a single feature in the model)?

Unfortunately the data is sensitive, but it's essentially a time series (with many other features that I discarded for the purpose of using only the time series and the target binary variable to forecast). If there were a multi-variable approach available, I would include the rest of these features.

However, X_train in my example above essentially has 2 columns ('date' where this is formatted as 0, 1, 2, etc and not as a datetime, and then 'y' which is the binary variable I am trying to forecast as '1' or '0')

Completely understand regarding the multi-variable forecasts. I would be happy to contribute to the implementation of this, where would be the best place to start?

**Comment by ourownstory** [comment]
Hi @mturk24,
I will try to unpack your questions and address them separately:
 
## Solution with existing code
#### Categorical variables: as target
As target variable (`'y'`), only singe-dimensional numerical variables have been implemented thus far. There is no support for categorical target variables. However, the case of a single categorical variable can be modelled as a binary numerical variable. 

If you had more categories (e.g. 3), you could only simplify it to a single dimensional numerical variable if they have an order and an attributable numeric value (e.g. `[0, 1.2, 5.1]`), which is rarely the case.

#### Categorical variables: as regressors
You should be able to use the majority of the additional features as inputs with our existing functionality.
Categorical inputs can be processed as regressors (lagged or future-known) or events (assumed to be future known), by creating a new variable for each category. 

You could write a simple helper function to do this for multi-categorical variables. It would not be the most elegant solution, but get the job done.

#### Non-datetime input data
Currently, the best approach is to simply map it to some random series of datestamps, as you seem to have been doing.
A helper function to do this might be useful, as dealing with datetimes can be messy.


## Creating a better Solution
#### Categorical variables: as regressors
This should be rather straight forward. When adding a regressor, an arg could specify it's a categorical variable, and how many categories there are. Then, the supplied category indexes `0,1,2,...` could be mapped to a one-hot representation `[1,0,0,...],...` (Pytorch has functions for this). Finally, you'd just have to make sure the `TimeNet` models it correctly (based on the additional arg in the `config`) and specify how to plot it.

#### Categorical variables: as target
This would be a bit more involved, and might actually be best implemented in conjunction with a multi-varible forecast, as it's essentially the same, just with a different loss function. 
Trend and seasonality could possibly be modelled for each category's logit. 
Plotting would likely have to be done for each category separately as a quick solution, but a joint plot for all categories might look better.

#### Non-datetime input data
Proper support for this has been on our roadmap for a while (tracked as issue #18 ).
I'll add some details on what that entails to the linked issue.

**Comment by mturk24** [comment]
@ourownstory Thank you for the detailed response, that is very helpful and I very much appreciate it! Do you have a simple example of code where you are using a binary categorical variable as a target (with only 2 values, 0 or 1) with just a time series to go along with it (as the feature space, technically)?

I still haven't gotten my code to run properly, and I just want to sanity check against something that satisfies this use case. I haven't found anything in the notebook examples or elsewhere to sanity check against.

**Comment by ourownstory** [comment]
@mturk24 
I am sorry, we do not have example code for this problem statement. 

Dataset:
To sanity check you could also just generate a timeseries artificially.
For a proper test, I looked and found these two datasets that could be modified to resemble your problem setting:

As is-inflow/ is-outflow
https://archive.ics.uci.edu/ml/datasets/CalIt2+Building+People+Counts

As game-happening or count>x
https://archive.ics.uci.edu/ml/datasets/Dodgers+Loop+Sensor

If you end up creating an example notebook it would be great if you could contribute it to the repo.


Finally, just to be sure I understand: You are not trying to do time-series classification, but forecasting, right?

**Reply by mturk24** [reply]
Ok thanks for the links! If I can get something working and find a compelling way to share, I'd love to make a notebook!

I'm trying to forecast the action column in the time series with the following example rows:
date:   action
1/1/76          0
1/2/76          1
1/3/76          1
etc.

Given the current implementation of neuralprophet, I thought it would be possible to do this. When observing my forecast of 'action', however, it's always in real-valued form (such as 0.726272 or 0.2232), despite following your suggestions and the documentation. I also made sure the column type is categorical for action.

**Reply by ourownstory** [reply]
> When observing my forecast of 'action', however, it's always in real-valued form (such as 0.726272 or 0.2232)

That is actually expected behavior, as the forecaster is set to output a numerical value. 
Until we properly support categorical forecasts, you could simply do this:
`yhat_cat = yhat > 0.5`

**Comment by ourownstory** [comment]
Converted this issue to a discussion and added issues #191, #192 and #193.
